src/pymoca/backends/casadi/api.py

```diff
@@ -308,19 +346,29 @@ def load_model(model_folder: str, model_name: str, compiler_options: Dict[str, s
             assert 1 == len({len(delay_expressions_raw), len(delay_durations_raw),
                 len(independent_delay_durations_raw)})
 
-            for i, expr in enumerate(delay_expressions_raw):
-                assert ca.depends_on(ca.MX(expr), all_symbols), \
-                    "Expression to be delayed can not be a constant"
+            all_symbols = [model.time,
+                           *model._symbols(model.states),
+                           *model._symbols(model.der_states),
+                           *model._symbols(model.alg_states),
+                           *model._symbols(model.inputs),
+                           *model._symbols(model.constants),
+                           *model._symbols(model.parameters)]
+
+            duration_dependencies = db['__delay_duration_dependent']
 
-                if ca.depends_on(ca.MX(delay_durations_raw[i]), all_symbols):
+            for i, expr in enumerate(delay_expressions_raw):
+                if duration_dependencies[i]:
                     dur = delay_durations_raw[i]
-                    false_deps = ca.vertcat(*[
-                        var for var in ca.symvar(dur) if not ca.depends_on(dur, var)])
-                    if false_deps.size1() > 0:
+
+                    deps = set(ca.symvar(dur))
+                    actual_deps = {all_symbols[j] for j in duration_dependencies[i]}
+                    false_deps = deps - actual_deps
+
+                    if false_deps:
                         [dur] = ca.substitute(
                             [dur],
-                            [false_deps],
-                            [ca.DM(np.full(false_deps.size1(), np.nan))])
+                            list(false_deps),
+                            [ca.repmat(np.nan, *d.size()) for d in false_deps])
```

Comment:
Not 100% sure, but I think that the numpy call is faster, and that you can use it without casting to DM. But maybe there is a good reason to use repmat?

Reply:
![image](https://user-images.githubusercontent.com/26876632/41355837-b1481106-6f22-11e8-835f-f2bdf505d3ac.png)

repmat is faster than what used to be there. Numpy arrays are 10x faster, but I think CasADi will convert to DM internally. And either way it's likely that `substitute` will be the bottleneck.

Reply:
agreed.